controller.go

```diff
@@ -64,40 +114,88 @@ func ServerLoop(listenPort int) {
 	port = listenPort
 
 	tcpAddr, err := net.ResolveTCPAddr("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		fmt.Printf("resolve tcp addr failed: %v", err)
+		return
+	}
 	listener, err := net.ListenTCP("tcp", tcpAddr)
-
-	ofc := NewOFController()
-	GetAppManager().RegistApplication(ofc)
-
 	if err != nil {
+		fmt.Printf("failed to listen: %v", err)
 		return
 	}
+	ofc := NewOFController()
+	ofc.RegisterHandler(h)
+	if err := ofc.Serve(listener); err != nil {
+		fmt.Printf("failed to serve: %v", err)
+		return
 
-	// wait for connect from switch
-	for {
-		conn, err := listener.AcceptTCP()
-		if err != nil {
-			return
-		}
-		go handleConnection(conn)
 	}
 }
 
+func (c *OFController) Close() error {
+	atomic.StoreInt32(&c.inShutdown, 1)
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	err := c.listener.Close()
+	for _, dp := range c.dps {
+		dp.Close()
+	}
+	return err
+}
+
+func (c *OFController) Shutdown(ctx context.Context) error {
+	atomic.StoreInt32(&c.inShutdown, 1)
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	err := c.listener.Close()
+	for _, dp := range c.dps {
+		dp.Shutdown()
+	}
+	for _, f := range c.onShutdown {
+		go f()
+	}
+	return err
+}
+
+func (c *OFController) isShuttingDown() bool {
+	return atomic.LoadInt32(&c.inShutdown) != 0
+}
+
+func (c *OFController) RegisterOnShutdown(f func()) {
+	c.mtx.Lock()
+	c.onShutdown = append(c.onShutdown, f)
+	c.mtx.Unlock()
+}
+
+func (c *OFController) RegisterHandler(h interface{}) {
+	c.mtx.Lock()
+	c.handlers = append(c.handlers, h)
+	c.mtx.Unlock()
+}
+
+func (c *OFController) GetHandler() []interface{} {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.handlers
```

Comment:
This looks potentially unsafe, it would allow iterating on the slice outside the lock while some other thread is modifying it inside the lock.

Reply:
Fixed.